File: translate-bubble/src/main/java/com/fortysevendeg/translace_bubble/ui/components/ContentView.java
package com.fortysevendeg.translace_bubble.ui.components;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.support.v4.view.GestureDetectorCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.*;
import android.widget.*;
import com.fortysevendeg.translace_bubble.R;

public class ContentView extends FrameLayout implements
        GestureDetector.OnGestureListener {

    public interface GestureListener {
        void onUp();
        void onDown();
        void onPrevious();
        void onNext();
    }

    private static final float SENSIBILITY = 80;

    private TextView original;

    private TextView translate;

    private TextView collapse;

    private TextView close;

    private ImageView options;

    private LinearLayout info;

    private LinearLayout buttons;

    private boolean showingInfo = true;

    private GestureDetectorCompat detector;

    private GestureListener gestureListener;

    public ContentView(Context context) {
        super(context);
        init();
    }

    public ContentView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public ContentView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
    }

    private void init() {
        ViewGroup view = (ViewGroup) LayoutInflater.from(getContext()).inflate(R.layout.content_view, null);
        original = (TextView) view.findViewById(R.id.original);
        translate = (TextView) view.findViewById(R.id.translate);
        collapse = (TextView) view.findViewById(R.id.collapse);
        close = (TextView) view.findViewById(R.id.close);
        info = (LinearLayout) view.findViewById(R.id.info_layout);
        buttons = (LinearLayout) view.findViewById(R.id.buttons_layout);
        options = (ImageView) view.findViewById(R.id.options);
        options.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                if (showingInfo) {
                    showButtons();
                } else {
                    showInfo();
                }
            }
        });
        showInfo();
        addView(view, new LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        detector = new GestureDetectorCompat(getContext(), this);
    }

    public void setGestureListener(GestureListener gestureListener) {
        this.gestureListener = gestureListener;
    }

    public void setTexts(String textOriginal, String textTranslate) {
        original.setText(textOriginal);
        translate.setText(textTranslate);
    }

    public void setListeners(OnClickListener collapseClickListener, OnClickListener closeClickListener) {
        collapse.setOnClickListener(collapseClickListener);
        close.setOnClickListener(closeClickListener);
    }

    private void showInfo() {
        showingInfo = true;
        info.setVisibility(VISIBLE);
        buttons.setVisibility(GONE);
        options.setImageResource(R.drawable.icon_options_light);
    }

    private void showButtons() {
        showingInfo = false;
        info.setVisibility(GONE);
        buttons.setVisibility(VISIBLE);
        options.setImageResource(R.drawable.icon_options_selected_light);
    }

    public void show() {
        showInfo();
        setVisibility(VISIBLE);
    }

    public void hide() {
        setVisibility(GONE);
    }

    public void collapse(final WindowManager.LayoutParams params, final WindowManager windowManager) {
        ValueAnimator animator = ValueAnimator.ofFloat(params.y, params.y + 100);
        animator.setDuration(100);
        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                float pos = (Float) animation.getAnimatedValue();
                params.alpha = 1 - (pos / 100);
                params.y = (int) pos;
                windowManager.updateViewLayout(ContentView.this, params);
            }
        });
        animator.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {

            }

            @Override
            public void onAnimationEnd(Animator animation) {
                hide();
                setAlpha(1);
                params.y = 0;
                params.alpha = 1;
                windowManager.updateViewLayout(ContentView.this, params);
            }

            @Override
            public void onAnimationCancel(Animator animation) {

            }

            @Override
            public void onAnimationRepeat(Animator animation) {

            }
        });
        animator.start();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        detector.onTouchEvent(event);
        return super.onTouchEvent(event);
    }

    @Override
    public boolean onDown(MotionEvent e) {
        return true;
    }

    @Override
    public void onShowPress(MotionEvent e) {

    }

    @Override
    public boolean onSingleTapUp(MotionEvent e) {
        return true;
    }

    @Override
    public boolean onScroll(MotionEvent e1, MotionEvent e2, float distanceX, float distanceY) {
        return true;
    }

    @Override
    public void onLongPress(MotionEvent e) {

    }

    @Override
    public boolean onFling(MotionEvent e1, MotionEvent e2, float velocityX, float velocityY) {
        if (gestureListener != null) {
            if ((e1.getY() - e2.getY()) > SENSIBILITY) {
                gestureListener.onUp();
            } else if ((e2.getY() - e1.getY()) > SENSIBILITY) {
                gestureListener.onDown();
            } else if ((e1.getX() - e2.getX()) > SENSIBILITY) {
                gestureListener.onNext();
            } else if ((e2.getX() - e1.getX()) > SENSIBILITY) {
                gestureListener.onPrevious();
            }
        }
        return true;
    }
}

File: src/main/scala/com/fortysevendeg/translatebubble/ui/components/ContentView.scala
/*
 * Copyright (C) 2015 47 Degrees, LLC http://47deg.com hello@47deg.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fortysevendeg.translatebubble.ui.components

import android.animation.{AnimatorListenerAdapter, Animator, ValueAnimator}
import android.content.Context
import android.text.TextUtils.TruncateAt
import android.util.AttributeSet
import android.view.ViewGroup.LayoutParams._
import android.view._
import android.widget.{FrameLayout, ImageView, LinearLayout, TextView}
import com.fortysevendeg.macroid.extras.ImageViewTweaks._
import com.fortysevendeg.macroid.extras.LinearLayoutTweaks._
import com.fortysevendeg.macroid.extras.ResourcesExtras._
import com.fortysevendeg.macroid.extras.TextTweaks._
import com.fortysevendeg.macroid.extras.ViewTweaks._
import com.fortysevendeg.translatebubble.R
import macroid.FullDsl._
import macroid._

import scala.language.postfixOps

class ContentView(context: Context, attrs: AttributeSet, defStyleAttr: Int)(implicit contextWrapper: ServiceContextWrapper)
    extends FrameLayout(context, attrs, defStyleAttr)
    with ContentViewLayout {

  def this(context: Context)(implicit contextWrapper: ServiceContextWrapper) = this(context, null, 0)

  def this(context: Context, attr: AttributeSet)(implicit contextWrapper: ServiceContextWrapper) = this(context, attr, 0)

  addView(layout, new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))

  private[this] def updateLayout(params: WindowManager.LayoutParams, windowManager: WindowManager) = Ui {
    windowManager.updateViewLayout(this, params)
  }

  def setTexts(langs: String, textOriginal: String, textTranslate: String): Ui[_] =
    (languages <~ tvText(langs)) ~ (original <~ tvText(textOriginal)) ~ (translate <~ tvText(textTranslate))

  def show(): Ui[_] = this <~ vVisible

  def hide(): Ui[_] = this <~ vInvisible

  def collapse(params: WindowManager.LayoutParams, windowManager: WindowManager): Ui[_] = Ui {
    val animator: ValueAnimator = ValueAnimator.ofFloat(0, 100)
    animator.setDuration(100)
    animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener {
      def onAnimationUpdate(animation: ValueAnimator) = {
        val pos: Float = animation.getAnimatedValue.asInstanceOf[Float]
        params.alpha = 1 - (pos / 100)
        runUi(updateLayout(params, windowManager))
      }
    })
    animator.addListener(new AnimatorListenerAdapter {
      override def onAnimationEnd(animation: Animator) = {
        params.alpha = 1
        runUi(hide() ~ updateLayout(params, windowManager))
      }
    })
    animator.start()
  }

  def changePositionIfIsNecessary(
      widthScreen: Int,
      heightScreen: Int,
      params: WindowManager.LayoutParams,
      windowManager: WindowManager): Ui[_] = {
    if (params.x + getWidth > widthScreen) {
      params.x = widthScreen - getWidth
    }
    if (params.y + getHeight > heightScreen) {
      params.y = heightScreen - getHeight
    }
    updateLayout(params, windowManager)
  }

}

trait ContentViewLayout {

  var languages = slot[TextView]

  var options = slot[ImageView]

  var original = slot[TextView]

  var translate = slot[TextView]

  def layout(implicit contextWrapper: ServiceContextWrapper) = getUi(
    l[LinearLayout](
      l[LinearLayout](
        w[TextView] <~ wire(languages) <~ tittleLanguagesStyle,
        w[ImageView] <~ wire(options) <~ optionsStyle
      ) <~ titleContentStyle,
      w[TextView] <~ wire(original) <~ originalTextStyle,
      w[ImageView] <~ lineStyle,
      w[TextView] <~ wire(translate) <~ translateTextStyle
    ) <~ rootStyle
  )

  def rootStyle(implicit contextWrapper: ContextWrapper): Tweak[LinearLayout] =
    vMatchParent +
        llVertical +
        vBackground(R.drawable.box)

  def titleContentStyle(implicit contextWrapper: ContextWrapper): Tweak[LinearLayout] =
    vMatchWidth +
        llHorizontal +
        llLayoutMargin(marginBottom = resGetDimensionPixelSize(R.dimen.margin_default)) +
        llGravity(Gravity.CENTER_VERTICAL)

  def tittleLanguagesStyle(implicit contextWrapper: ContextWrapper): Tweak[TextView] =
    llWrapWeightHorizontal +
        vPadding(paddingBottom = resGetDimensionPixelSize(R.dimen.margin_default)) +
        tvColorResource(R.color.languages_content_light) +
        tvSizeResource(R.dimen.languages_content) +
        tvLines(1) +
        tvEllipsize(TruncateAt.END) +
        tvBold

  def optionsStyle(implicit contextWrapper: ContextWrapper): Tweak[ImageView] =
    vWrapContent +
        llLayoutGravity(Gravity.TOP | Gravity.RIGHT) +
        ivSrc(R.drawable.box_icon_close)

  def originalTextStyle(implicit contextWrapper: ContextWrapper): Tweak[TextView] =
    vMatchWidth +
        tvColorResource(R.color.title_content_light) +
        tvSizeResource(R.dimen.text_content) +
        tvLines(1) +
        tvEllipsize(TruncateAt.END)

  def lineStyle(implicit contextWrapper: ContextWrapper): Tweak[ImageView] = {
    val padding = resGetDimensionPixelSize(R.dimen.padding_default)
    lp[ViewGroup](MATCH_PARENT, resGetDimensionPixelSize(R.dimen.line_stroke)) +
        llLayoutMargin(marginTop = padding, marginBottom = padding) +
        vBackgroundColorResource(R.color.line_content)
  }

  def translateTextStyle(implicit contextWrapper: ContextWrapper): Tweak[TextView] =
    llMatchWeightVertical +
        tvColorResource(R.color.text_content_light) +
        tvSizeResource(R.dimen.text_content) +
        tvEllipsize(TruncateAt.END)

}
